Document DynamicDependencyAttribute

sbomer opened the pull request:
Adds documentation for DynamicDependencyAttribute in the instructions for preparing libraries for trimming.

sbomer committed:
Document DynamicDependencyAttribute

sbomer committed:
Fix bare URL and adjust wording

vitek-karas commented on docs/core/deploying/prepare-libraries-for-trimming.md:
```diff
@@ -177,6 +177,43 @@ static void InitializeTypeField()
+### DynamicDependency
+
+This attribute can be used to indicate that a member has a dynamic dependency on other members. This results in the referenced members being kept whenever the member with the attribute is kept, but doesn't silence warnings on its own.
+This can be used together with `UnconditionalSuppressMessageAttribute` to fix some analysis warnings.
```
I would add a warning like:
Usage of `DynamicDependencyAttribute` should be limited to cases where no other approach is usable. Trimming tools can't validate that the dependency specified by the attribute will be enough to make the application work, so it's very likely trimming will produce warnings even if the usage of the attribute fully covers the use case. If such warnings are suppressed (via `UnconditionalSuppressMessageAttribute`) trimmer will also not validate that potential future changes in the code are correctly reflected in the usage of `DynamicDependencyAttribute`.

vitek-karas replied:
In short - we don't want people to use this - unless they absolutely have to.

sbomer replied:
I moved this down to the "advanced" section and added some warnings - PTAL.

vitek-karas reviewed (approved):
Other than the warning which I think should be added, this looks great. Thanks!

eerhardt commented on docs/core/deploying/prepare-libraries-for-trimming.md:
```diff
@@ -177,6 +177,43 @@ static void InitializeTypeField()
+[DynamicDependency("Helper", "MyType", "MyAssembly")]
+static void RunHelper()
+{
+    var helper = Assembly.Load ("MyAssembly").GetType("MyType").GetMethod("Helper");
```
```suggestion
    var helper = Assembly.Load("MyAssembly").GetType("MyType").GetMethod("Helper");
```

eerhardt commented on docs/core/deploying/prepare-libraries-for-trimming.md:
```diff
@@ -177,6 +177,43 @@ static void InitializeTypeField()
+
+The attribute specifies the members to keep via a `string` or via `DynamicallyAccessedMemberTypes`. The type and assembly are either implicit in the attribute context, or explicitly specified in the attribute (by `Type`, or by `string`s for the type and assembly name).
+
+The type and member strings use the format described at C# documentation comment [format](https://github.com/dotnet/csharplang/blob/master/spec/documentation-comments.md#id-string-format), without the member prefix. The member string should not include the name of the declaring type, and may omit parameters to keep all members of the specified name. Some examples of the format follow:
```
(nit) Is this format doc'd on docs.microsoft.com at all? I wonder how much linking we should be doing from docs.microsoft.com to github docs. Especially considering the branch `master` might be renamed at some point. Or the doc could be moved.

sbomer replied:
Turns out it does exist on docs.microsoft.com - I fixed the link. Thanks!

eerhardt commented on docs/core/deploying/prepare-libraries-for-trimming.md:
```diff
@@ -177,6 +177,43 @@ static void InitializeTypeField()
+[DynamicDependency("MemberName", "Namespace.ContainingType.NestedType", "Assembly")]
+// generics
+[DynamicDependency("GenericMethodName`1")]
+[DynamicDependency("GenericMethod``2(``0,``1)")]
```
Why the double ``` `` ``` here?

sbomer replied:
All of the generic method ones were supposed to use double ` `` ` actually - I fixed `GenericMethodName` to do the same. That's just the format for generic method parameters (generic parameters on types use one, generic parameters on methods use two).

eerhardt reviewed (approved):
Thanks!

sbomer committed:
Merge remote-tracking branch 'origin/main' into dynamicDependency

sbomer committed:
PR feedback

- Fix generic method parameter format
- Add warning text
- Move DynamicDependency section down
- Fix whitespace in code snippet
